### PR TITLE
feat: revamp plugin architecture

### PR DESCRIPTION
**Pull request by louis-jan**
- [x] Revamp plugin management -> plugin as class, lifecycle
- [x] Filesystem apis (core/fs)
- [x] Inference request should not care about db / stored conversations, whatever it need, pass as input. cc @namchuai 
- [x] No continuous syncing, delayed sync instead, app works with its in-memory states 
- [x] Revamp conversation plugin (json / markdown)
- [x] Revamp inference plugin
- [x] Revamp monitoring plugin 
- [x] Revamp model plugin 

Related issue: #403 

Fixes: #483

### CODE REVIEW

File: web/containers/Providers/index.tsx [hunk 1]
```diff
@@ -34,6 +36,8 @@ const Providers = (props: PropsWithChildren) => {
 
     // Register all active plugins with their activation points
     await plugins.registerActive()
```
Comment (louis-jan) [inline web/containers/Providers/index.tsx:38]: Will remove when all plugins are fully ported

File: web/helpers/EventHandler.tsx [hunk 2]
```diff
@@ -75,27 +89,54 @@ export default function EventHandler({ children }: { children: ReactNode }) {
...
+    if (!convo) {
+      return
+    }
+    pluginManager
+      .get<ConversationalPlugin>(PluginType.Conversational)
+      ?.saveConversation({
+        ...convo,
+        _id: convo._id ?? '',
+        name: convo.name ?? '',
+        messages: (messagesRef.current ?? [])[convo._id].map<Message>(
```
Comment (louis-jan) [inline web/helpers/EventHandler.tsx:128]: TODO: add bot setting

File: web/helpers/EventHandler.tsx [hunk 3]
```diff
@@ -75,27 +89,54 @@ export default function EventHandler({ children }: { children: ReactNode }) {
...
 
   async function handleMessageResponseFinished(
     messageResponse: NewMessageResponse
   ) {
-    if (!messageResponse.conversationId) return
-    console.debug('handleMessageResponseFinished', messageResponse)
+    if (!messageResponse.conversationId || !convoRef.current) return
     updateConvWaiting(messageResponse.conversationId, false)
+
+    // TODO: test persistence
```
Comment (louis-jan) [inline web/helpers/EventHandler.tsx:115]: Clean

File: plugins/inference-plugin/index.ts [hunk 4]
```diff
@@ -4,31 +4,29 @@ import {
...
     const requestBody = JSON.stringify({
       messages: recentMessages,
       stream: true,
       model: "gpt-3.5-turbo",
-      max_tokens: bot?.maxTokens ?? 2048,
-      frequency_penalty: bot?.frequencyPenalty ?? 0,
-      presence_penalty: bot?.presencePenalty ?? 0,
-      temperature: bot?.customTemperature ?? 0,
+      max_tokens: 2048,
+      // TODO: Enable back when character is available
```
Comment (louis-jan) [inline plugins/inference-plugin/index.ts:26]: Add back when combined with #526

**Comment by louis-jan** [comment]
@namchuai, I removed these. The app should work with its state only and perform delayed sync (periodically/checkpoints). Continuous updating seems to be causing poor performance

getConversationById(id: string): Promise<any> 
createConversation(conversation: any): Promise<number | undefined> 
updateConversation(conversation: any): Promise<void>
createMessage(message: any): Promise<number | undefined>
updateMessage(message: any): Promise<void>
getConversationMessages(conversationId: any)

**Review by namchuai — APPROVED** [review]

**Pull request merged**